feat: our feature doc (#55)
elastic/apm-aws-lambda: apm-lambda-extension/docs/aws-lambda-extension.asciidoc

[[aws-lambda-extension]]
= AWS Lambda Extension (Experimental)

> Elastic's APM Language Agents are able to instrument AWS Lambda functions via an AWS Lambda Extension.

Reviewer: ```suggestion
Elastic's APM Agents instrument AWS Lambda functions via an AWS Lambda Extension.
```

[[aws-lambda-arch]]
== Extension Architecture

> Normally, during the execution of a Lambda function, there's only a single language process running in the AWS Lambda execution environment.  However, with an AWS Lambda Extension, Lambda users can run a _second_ process along side their main service/application process.

Reviewer: ```suggestion
Normally, during the execution of a Lambda function, there's only a single language process running in the AWS Lambda execution environment.  However, with an AWS Lambda Extension, Lambda users can run a _second_ process alongside their main service/application process.
```

// TODO: replace below with a diagram diagram

[source,txt]
----
    [Lamnda Function/Service] -> [Lambda Extension] -> [APM Server]
----

> By using a custom built AWS Lambda Extension, Elastic APM Agents can send data to a locally running Lambda Extension process, and that process will forward data on to APM Server.  The Lambda Extension ensures that any latency between the Lambda Function and the AWS Server instance will not cause latency in the Lambda Function/Service itself.

Reviewer: ```suggestion
By using a custom-built AWS Lambda Extension, Elastic APM Agents can send data to a locally running Lambda Extension process, and that process will forward data on to APM Server.  The Lambda Extension ensures that any latency between the Lambda Function and the AWS Server instance will not cause latency in the Lambda Function/Service itself.
```

Reviewer: It might make sense to move this paragraph above the diagram.

[[aws-lambda-install]]

> == Installing the Lambda Extension

Reviewer: We don't have to add it now, but I'd like to eventually see an overview of all of the steps required at the start of this guide.

> Elastic offers an installer for adding the Lambda Extension to your Lambda functions.  This installer will

Reviewer: ```suggestion
Elastic offers an installer for adding the Lambda Extension to your Lambda functions.  This installer will:
```

1. Compile the Lambda Extension from source (written in go)
2. Publish the Lambda Extension as a layer

> 3. Configure a named Lambda Function with the just published layer

Reviewer: Should Function be capitalized here?

> 4. Configure the required environmental variables.

Reviewer: All other OL items don't have a period. Let's remove it from number four to be consistent.
```suggestion
4. Configure the required environmental variables
```

> If you'd rather manually install and configuration the Lambda Extension, see the Manually Installing section below.

Reviewer: Where is this section?

Author: Looks like it was renamed to _Configuring the Agent and Lambda Function handler_ -- will update the text to reflect that.

> The installer is distributed avia GitHub as a Node.js project.  Once you've downloaded the project and installed a version of Node.js run the following two commands.

Reviewer 2: ```suggestion
The installer is distributed via GitHub as a Node.js project.  Once you've downloaded the project and installed a version of Node.js run the following two commands.
```

[source,shell]
----
    $ npm install   # installs the dependencies for the cli
    $ ./elastic-lambda.js install
----

The `elastic-lambda.js` command assumes you have a `install.yaml` file configured.  There's a sample of this file distributed with the repository.  To use it, just cope the file and edit its contents.

[source,shell]
----
    $ cp install.yaml install.yaml.dist
----

> **Important**: The installer assumes your local environment is configured to authenticate against the AWS using Amazon's standard environment variables.  Depending on your authentication method, this may look something like the following

Reviewer: is "the AWS" correct here?

[source,shell]
----
    $ AWS_DEFAULT_REGION=us-west-2 \
      AWS_ACCESS_KEY_ID=AKIAZEDJODE3B3UMDAKX \
      AWS_SECRET_ACCESS_KEY=hmE7n1gfiyXzgwOQu2bxOA92HrVVWh8WG \
      ./elastic-lambda.js install
----

The meaning of each `install.yaml` configuration field is as follows.

=== `layer_name`

> The is the name the compiler will use for your AWS Layer.  The default, `apm-lambda-extension`, should work for most scenarios.

Reviewer 2: ```suggestion
This is the name the compiler will use for your AWS Layer.  The default, `apm-lambda-extension`, should work for most scenarios.
```

Thinking about the default name, I think `elastic-apm-extension` might be better. Having `lambda` in the name seems redundant as the context the extension is used is always lambda. To disambiguate which apm extension that is, adding `elastic` could be helpful.

=== `function_name`

The name of your Lambda Function.  The installer will use this to configure the correct lambda function.  This must be the name of a function that already exists.

=== `lambda_env`

> The installet will use the key/value pairs in this section of the configuration file to add an environmental variable configuration to your Lambda function.  The provided variable are those required to make the extension function correctly.

Reviewer 2: ```suggestion
The installer will use the key/value pairs in this section of the configuration file to add environment variables to your Lambda function.  The provided variables are those required to make the extension work correctly.
```

Function is a overloaded term in this context.

==== `ELASTIC_APM_LOG_LEVEL`

> The log level for the APM Agent.

Reviewer 2: How can the log level of the extension be configured?

Author: Good question.  At this time it _can't_ be configured.  I've open an issue to cover this: https://github.com/elastic/apm-aws-lambda/issues/60

==== `ELASTIC_APM_SECRET_TOKEN`

The APM secret token.  The extension will use this when communicating with APM Server.

==== `ELASTIC_APM_API_KEY`

An alternative authentication method to the secret token.  The extension will use this when communicating with APM Server.

==== `ELASTIC_APM_SERVER_URL`

Your APM Server URL.  This is the final destination for your data.

==== `ELASTIC_APM_SERVICE_NAME`

> The configured name of your application or service.  The APM Agent will use this value when reporting data to APM Server.

Reviewer 2: ```suggestion
The configured name of your application or service.  The APM Agent will use this value when reporting data to APM Server.
If unset, the APM Agent will automatically set the value based on `AWS_LAMBDA_FUNCTION_NAME` or `context.functionName`.
```

> ==== `ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS`

Reviewer 2: Not quite sure where the timeout applies. Is the data receiver the extension or APM Server? Maybe consider renaming to `ELASTIC_APM_EXTENSION_TIMEOUT_SECONDS`?

Author: Another good question.  The HTTP Server that the extension starts (in order to receive data from a language agents) requires a timeout value.  That value is used here: https://github.com/elastic/apm-aws-lambda/blob/63afa0a37193f0e97f3bed185138bf492eb7f6a2/apm-lambda-extension/extension/http_server.go#L59-L60

I've opened an issue to reconsider the name for this env value, as well as whether we actually need it: https://github.com/elastic/apm-aws-lambda/issues/61

The timeout value, in seconds, for the Lambda Extension's server.

== Configuring the Agent and Lambda Function handler

Once you've installed the extension, there's one last step to take. You'll need to wrap the Lambda Function handler.

[[aws-lambda-nodejs]]
=== Node.js

In Node.js, you wrap a Lambda Function handler using the following syntax.

[source,js]
----
const apm = require('elastic-apm-node').start({/*...*/})
exports.handler = apm.lambda(async function handler (event, context) {
    const response = {
      statusCode: 200,
      body: "hello new async."
    };
    return response
})
----

[[aws-lambda-python]]
=== Python

In Python, you wrap a Lambda Function handler using the following syntax.

[source,python]
----
from elasticapm import capture_serverless
@capture_serverless()
def handler(event, context):
    return {"statusCode": r.status_code, "body": "Success!"}
----

[[aws-lambda-java]]
=== Java

> In Java, you wrap a Lambda Function handler using the following syntax.

Reviewer 3: We need to decide whether we want to create an additional Java-agent layer (so users would configure two layers to be used for Java), or whether we rather would create a dedicated layer for Java that would contain both, the java agent and the extension.

// TODO: replace below with a diagram diagram

[source,txt]
----
    [Lamnda Function/Service] -> [Lambda Extension] -> [APM Server]
----

By using a custom built AWS Lambda Extension, Elastic APM Agents can send data to a locally running Lambda Extension process, and that process will forward data on to APM Server.  The Lambda Extension ensures that any latency between the Lambda Function and the AWS Server instance will not cause latency in the Lambda Function/Service itself.

[[aws-lambda-install]]
== Installing the Lambda Extension

Elastic offers an installer for adding the Lambda Extension to your Lambda functions.  This installer will:

1. Compile the Lambda Extension from source (written in go)
2. Publish the Lambda Extension as a layer
3. Configure a named Lambda Function with the just published layer
4. Configure the required environmental variables

If you'd rather manually install and configuration the Lambda Extension, see the Manually Installing section below.

> The installer is distributed via GitHub as a Node.js project.  Once you've downloaded the project and installed a version of Node.js run the following two commands.

Reviewer 4: Maybe specify what it means to "download the project". I.e. clone the repo?

Also, you need to cd into the `/cli` directory before you can run these commands. Might be worth mentioning.

2. Click on the _Configuration_ tab
3. Click on _Environment variables_
4. Add the necessary variables.

=== The Necessary Variables

==== `ELASTIC_APM_CENTRAL_CONFIG`

The `ELASTIC_APM_CENTRAL_CONFIG` value _must_ be set to `false`. Central configuration does not work in a Lambda environment, and having this on will negatively impact the performance of your Lambda function.

==== `ELASTIC_APM_CLOUD_PROVIDER`

The `ELASTIC_APM_CLOUD_PROVIDER` value _must_ be set to `none`.  Amazon's Cloud Metadata APIs are not available in an AWS Lambda environment, and attempting to fetch this data will negatively impact the performance of your Lambda function.

==== `ELASTIC_APM_LAMBDA_APM_SERVER`

The `ELASTIC_APM_LAMBDA_APM_SERVER` controls where the Lambda extension will ship data.  This should be the URL of the final APM Server destination for your telemetry.

==== `ELASTIC_APM_SECRET_TOKEN` or `ELASTIC_APM_API_KEY`

Either `ELASTIC_APM_API_KEY` or `ELASTIC_APM_SECRET_TOKEN` needs to be set.  This controls the authentication method that extension uses when sending data to the URL configured via `ELASTIC_APM_LAMBDA_APM_SERVER`.

==== `ELASTIC_APM_SERVER_URL`

> This _must_ be configured to the value `http://localhost:8200`.  This configuration field controls where your APM Agent sends data.  The extension listens for data `localhost:8200`.

Reviewer 3: ```suggestion
This _must_ be configured to the value `http://localhost:8200`.  This configuration field controls where your APM Agent sends data.  The extension listens for data on `localhost:8200`.
```
